main/eval/src/mill/eval/EvaluatorCore.scala

```diff
@@ -66,93 +66,117 @@ private[mill] trait EvaluatorCore extends GroupEvaluator {
+    def evaluateTerminals(terminals: Seq[Terminal], contextLoggerMsg: Int => String)(implicit
+        ec: ExecutionContext
+    ) = {
+      // We walk the task graph in topological order and schedule the futures
+      // to run asynchronously. During this walk, we store the scheduled futures
+      // in a dictionary. When scheduling each future, we are guaranteed that the
+      // necessary upstream futures will have already been scheduled and stored,
+      // due to the topological order of traversal.
+      for (terminal <- terminals) {
+        val deps = interGroupDeps(terminal)
+        futures(terminal) = Future.sequence(deps.map(futures)).map { upstreamValues =>
+          if (failed.get()) None
+          else {
+            val upstreamResults = upstreamValues
+              .iterator
+              .flatMap(_.iterator.flatMap(_.newResults))
+              .toMap
+
+            val startTime = System.nanoTime() / 1000
+            val threadId = threadNumberer.getThreadId(Thread.currentThread())
+            val counterMsg = s"${count.getAndIncrement()}/${terminals.size}"
+            val contextLogger = PrefixLogger(
+              out = logger,
+              context = contextLoggerMsg(threadId),
+              tickerContext = GroupEvaluator.dynamicTickerPrefix.value
+            )
+
+            val res = evaluateGroupCached(
+              terminal = terminal,
+              group = sortedGroups.lookupKey(terminal),
+              results = upstreamResults,
+              counterMsg = counterMsg,
+              zincProblemReporter = reporter,
+              testReporter = testReporter,
+              logger = contextLogger
             )
-          )
 
-          Some(res)
+            if (failFast && res.newResults.values.exists(_.result.asSuccess.isEmpty))
+              failed.set(true)
+
+            val endTime = System.nanoTime() / 1000
+
+            val duration = endTime - startTime
+
+            chromeProfileLogger.log(
+              task = Terminal.printTerm(terminal),
+              cat = "job",
+              startTime = startTime,
+              duration = duration,
+              threadId = threadNumberer.getThreadId(Thread.currentThread()),
+              cached = res.cached
+            )
+
+            profileLogger.log(
+              ProfileLogger.Timing(
+                terminal.render,
+                (duration / 1000).toInt,
+                res.cached,
+                deps.map(_.render),
+                res.inputsHash,
+                res.previousInputsHash
+              )
+            )
+
+            Some(res)
+          }
         }
       }
     }
 
-    val finishedOptsMap = terminals
+    val tasks0 = terminals0.filter {
+      case Terminal.Labelled(c: Command[_], _) => false
+      case _ => true
+    }
+
+    val (_, tasksTransitive0) = Plan.plan(Agg.from(tasks0.map(_.task)))
+
+    val tasksTransitive = tasksTransitive0.toSet
+    val (tasks, commands) = terminals0.partition {
```

Comment:
Can we either add some comment or rename `commands` to make it more obvious, it's only those commands that don't have downstream dependencies?

Reply:
I renamed it `leafCommands`. Ideally I would want `terminalCommands`, but `Terminal` already has another meaning in this code

Reply:
So, IIUC, we don't evaluate all command serially, only those, that don't have downstream dependencies. That's a bit odd, as I was hoping, we could the user some guarantee, that commands are always executed serially. 

Now, that I think about it, we probably only want commands, that are interactive to run serially. But we currently can't determite, if a tasks wants to be interactive.

Reply:
Not just interactive commands, there's probably lots of commands to do side effects, tests that are not well behaved (e.g. writing to shared paths on disk), and commands that print stuff (and probably don't want the `[#threadid]` prefix). Those are all solvable, but I think "only commands at the end of the evaluation graph run serially" is a reasonable compromise that gets us most of the way there

Reply:
Understood. 

I don't know how often targets depend on commands, but I could imagine, it would be fine to just run those downstream dependencies serially too. So we could make this guarantee of all commands without implementing multiple parallel/serial blocks.